rfcs/45-stageless.md

```diff
@@ -0,0 +1,1078 @@
+5. `apply_player_movement` is before `apply_velocity`
+   1. Real: these systems conflict on `Velocity`
+
+The problem here is that there's no direct link between `determine_player_movement` and `apply_player_movement`: they are ambiguous under Model 2, and the player may move according to stale input!
+
+Under the transitive Model 1, additional if-needed constraints are created between `determine_player_movement` and `apply_player_movement` via the constraint chain (1, 2, 4).
+Then, the inferred constraint between `determine_player_movement` and `apply_player_movement` is converted into a strict ordering constraint, as the systems are incompatible.
+Finally, all redundant strict ordering constraints (such as between `collision_detection` and `apply_velocity`) are discarded for performance reasons.
+
+Under Model 2, the behavior is much simpler: `determine_player_movement` and `apply_player_movement` are ambiguous.
+
+Model 1 preserves the "logical" behavior, but in a very implicit fashion that is hard to reason about.
+Model 2 simply causes the user's logic to break.
+
+By adding powerful (and prominent) tools for detecting spurious ordering constraints and execution order ambiguities, we can follow the expected (and convenient) transitive property while helping users quash strange, fragile bugs as soon as they're introduced.
+
+### What syntax should we use for collections of systems?
+
+When using convenience methods like `add_system_chain` and `add_system_set`, we need to be able to refer to conveniently refer to collections of systems.
+This will become increasingly important as more complex graph-builder APIs are added.
+
+In summary:
+
+- array syntax: optimal but literally impossible
+- tuple syntax (winner!): magic but pretty
+- builder syntax: way too verbose for a pattern that's supposed to increase convenience
+- `vec!`-style macro: not any less magic than tuple syntax, not as pretty
+
+Below, we examine the the options by example.
+
+#### Arrays
+
+```rust
+.add_system_chain([compute_attack, 
+                   compute_defense,
+                   check_for_crits,
+                   compute_damage,
+                   deal_damage,
+                   check_for_death]
+                   .label(GameLabel::Combat))
+
+.add_systems([run, jump].after(CoreLabel::Input))
+```
+
+Very pretty. Doesn't work because types are heterogenous though.
+
+#### Tuples
+
+```rust
+.add_system_chain((compute_attack, 
+                   compute_defense,
+                   check_for_crits,
+                   compute_damage,
+                   deal_damage,
+                   check_for_death)
+                   .label(GameLabel::Combat))
+
+.add_systems((run, jump).after(CoreLabel::Input))
+```
+
+Also pretty! Mildly cursed, but we use this flavor of cursed type system magic already for bundles.
+
+Requires another invocation of the `all_tuples` macro to generate impls.
+
+#### Builder pattern
+
+This was the strategy we used for `SystemSet`.

```

Comment:
While I like the look of the tuple one better, the builder pattern might be a better solution for handling configuration.

```rust
let mut my_group = SystemGroup:::new().with(system_a).with(system_b);

#[cfg(feature = "do_optional_thing")]
my_group.with(optional_system);

my_group.with(system_c);
```

Reply:
Under the hood, I think we need to have a `SystemGroup` type anyways, with the corresponding API.

Reply:
Clarified, thanks :)